Jenkinsfile

```diff
@@ -0,0 +1,165 @@
+!groovy
+
+JENKINS_CREDS = [
+    [
+        $class: 'UsernamePasswordMultiBinding',
+        credentialsId: 'jenkins-docker-credentials',
+        passwordVariable: 'DOCKER_PASSWORD',
+        usernameVariable: 'DOCKER_USERNAME'
+    ]
+]
+
+def buildDockerImage(commit, version, repo) {
+    sh """
+        docker build --rm \\
+            --build-arg git_commit=${commit} \\
+            --build-arg version=${version} \\
+            --build-arg descriptive_version=${version} \\
+            -t ${repo} \\
+            .
+    """
+
+    // Retrieve the image SHA.
+    imageSha = sh(returnStdout: true, script: "docker inspect -f '{{ .Config.Image }}' ${repo}").trim()
+    echo imageSha
+
+    // Record the image SHA.
+    shaFile = "${repo}.docker-image-sha"
+    writeFile(file: "${shaFile}", text: "${imageSha}")
+    fingerprint "${shaFile}"
+}
+
+def testGoService(containerName, repo, packagesToTest) {
+    sh """
+        docker run --rm \\
+            --name ${containerName} \\
+            --entrypoint 'go' \\
+            ${repo} \\
+            test ${packagesToTest}
+    """
```

Comment:
With other go stuff that has tests, I've set it up to use `go-junit-report` and Jenkins' integrated JUnit publishing stuff so we get pretty graphs; see e.g. https://github.com/cyverse-de/saved-searches/blob/8220dd07ec488e254faf607d56ca058eebf7d2b7/Jenkinsfile#L31

In `querydsl` that I've been working on, I've also been more dramatic/blunt-force about which packages to test, and just `cd`ing to the directory and then using `go test -v ./...` to have it test everything: https://github.com/cyverse-de/querydsl/blob/3c37f12f07da0f165d64d82a4c7c3cd7ec7fdbc4/Dockerfile#L14 -- this might or might not be appealing to you, but I thought I'd let you know it's an option in case you didn't! (querydsl, since it's a library, doesn't publish the docker image, so I have the go-junit-report stuff in the Dockerfile, but the same approach should be totally feasible in the jenkinsfile too of course)

Reply:
I considered testing everything, but I didn't want to include the vendor directory.

Reply:
Oh, d'oh. I forgot about that, since it didn't affect me in `querydsl`. Makes total sense.

I'd still recommend using `go-junit-report` though!

Reply:
Yeah, that might be useful.